File: 0974/solution.ts
function subarraysDivByK(nums: number[], k: number): number {
  const counts = new Map<number, number>();
  counts.set(0, 1);
  let total = 0;
  let res = 0;
  for (const num of nums) {
    total += num;
    let remainder = total % k;
    if (remainder < 0) {
      remainder += k;
    }
    const currCount = counts.get(remainder) || 0;
    res += currCount;
    counts.set(remainder, currCount + 1);
  }
  return res;
}

File: 0974/solution.java
import java.util.HashMap;
import java.util.Map;

class Solution {
   public int subarraysDivByK(int[] nums, int k) {
      Map<Integer, Integer> counts = new HashMap<>();
      counts.put(0, 1);
      int total = 0;
      int res = 0;
      for (int num : nums) {
         total += num;
         int remainder = total % k;
         if (remainder < 0) {
            remainder += k;
         }
         int currCount = counts.getOrDefault(remainder, 0);
         res += currCount;
         counts.put(remainder, currCount + 1);
      }
      return res;
   }
}

// time O(n)
// space O(min(k, n))